logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientModel.kt

```diff
@@ -0,0 +1,72 @@
+package com.wire.kalium.logic.data.client
+
+import com.wire.kalium.logic.data.location.Location
+import com.wire.kalium.logic.data.prekey.PreKey
+import com.wire.kalium.network.api.user.client.ClientTypeDTO
+
+
+data class RegisterClientParam(
+    val password: String,
+    val preKeys: List<PreKey>,
+    val lastKey: PreKey,
+    val deviceType: DeviceType,
+    val type: ClientType, // 'temporary', 'permanent', 'legalhold'
+    val label: String,
+    val capabilities: List<ClientCapability>?,
+    val model: String
+)
+
+data class Client(
+    val clientId: String,
+    val type: ClientType,
+    val registrationTime: String, // yyyy-mm-ddThh:MM:ss.qqq
+    val location: Location?,
+    val deviceType: DeviceType?,
+    val label: String?,
+    val cookie: String?,
+    val capabilities: Capabilities?,
+    val model: String?
+)
+
+data class Capabilities(
+    val capabilities: List<ClientCapability>
+)
+
+enum class ClientType {
+    Temporary,
+    Permanent,
+    LegalHold;
+
+    override fun toString(): String {
+        return this.name.lowercase()
+    }
+
+
+
+    fun ClientTypeDTO.toClientType(): ClientType {
+        return when(this) {
+            ClientTypeDTO.Temporary -> Temporary
+            ClientTypeDTO.Permanent -> Permanent
+            ClientTypeDTO.LegalHold -> LegalHold
+        }
+    }
+}
+
+enum class DeviceType {
+    Phone,
+    Tablet,
+    Desktop,
+    LegalHold;
+
+    override fun toString(): String {
+        return this.name.lowercase()
+    }
+}
+
+enum class ClientCapability {
+    LegalHoldImplicitConsent {
+        override fun toString(): String {
```

Comment:
In general, not specifically in this case, I'm confused about the use of `toString`. It this some sort of string representation of a client capability? 

In Swift for example there isn't a `toString` method. But rather:

- `description: String`. Used for [describing an object](https://developer.apple.com/documentation/swift/customstringconvertible) as a String and called when String interpolation takes place
- `debugDescription: String`. [This is self explanatory](https://developer.apple.com/documentation/swift/customdebugstringconvertible) 
- `as? String` to cast something, to String. The cast will fail if the object is not a String.

There are also other ways that achieve [slightly](https://developer.apple.com/documentation/swift/losslessstringconvertible/) different [things]((https://developer.apple.com/documentation/swift/expressiblebystringliteral))

None of these to, are "magically" added to any object.

Reply:
@tmspzz `toString` is kind of in a no-man's-land. It is something Kotlin inherited from Java and it may be considered harmful for some people.

Following Kotlin's documentation on it:
> Returns a string representation of the object. Can be called with a null receiver, in which case it returns the string "null".

There's no specifically-set purpose for it. But I'd say it's closer to swift's `description`, and some people will use it in a `debugDescription` way. 

The default implementation of `toString()` in Kotlin:
* For classes will return something like: `MyClass@1eed786` where `1eed786` is the memory address where the object is stored. Of course you can inherit the `toString` from parent classes.
* For primitives: their value,
* For `enum`s: their name as written in code.

The `as? String` is the same in Kotlin as in Swift. It will crash if the thing you're trying to cast isn't a String.

Reply:
For this case, I understand that `toString` is being overriden to match the API and work the (de)serialization, which for an `enum` created solely for this, seems reasonable.

BUT. I think making it clear why the `toString` is being overriten is always great, and overwriting it for each possible enum value can be too verbose.

Instead of:

```kotlin
enum class Make {
    AlfaRomeo {
        override fun toString(): String = "alfa-romeo"
    },
    Ferrari {
        override fun toString(): String = "ferrari"
    },
    Fiat {
        override fun toString(): String = "fiat"
    }
}
```

This can be done:

```kotlin
@Serializable
enum class Make {
    @SerialName("alfa-romeo") AlfaRomeo,
    @SerialName("ferrari") Ferrari,
    @SerialName("fiat") Fiat
}
```

No need to alter how the default `toString()` works, less code, and makes it clear why the different name exists.

Reply:
`toString` definitely harmful in my opinion

Reply:
using only '''@SerialName("alfa-romeo")''' will not work in our serlization tests, since when creating the json string kinda manually e.g. ''' "make": "${Make.AlfaRomeo}" ''' the result string will be ''' "make": "AlfaRomeo" ''' and it will differ from the  '''@SerialName("alfa-romeo")'''. 
so we need the '''toString''' of an enum, which is it's name in kotlin, to match the '''@SerialName'''.

Reply:
I guess I shot myself on the foot a few of weeks ago when I did this `JsonProvider` thing 😿 

Definitely an improvement point to think about regarding `JsonProvider`.

Reply:
we can remove the `toString` from the enums and change the json tests to  to something like `"make": "alfa-romeo"`